`make apidocs` assumes `natcap.invest` has been built from source

`make apidocs` assumes that `natcap.invest` has been built from source and does not ensure that it has.

The workaround is to `rm -r build dist/*.whl; make install; make apidocs` in order to get the build files into the expected places.

To reproduce:

* `rm -r build dist`
* Activate your conda environment
* `make apidocs`
* Observe error:
```python
mkdir -p dist
make: Circular dist/apidocs <- dist/apidocs dependency dropped.
# -a: always build all files
python3 -m sphinx -a -b html \
		-d build/sphinx/apidocs \
		doc/api-docs build/sphinx/apidocs
Running Sphinx v4.0.1

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/Users/jdouglass/workspace/natcap/invest/env38/lib/python3.8/site-packages/sphinx/config.py", line 323, in eval_config_file
    exec(code, namespace)
  File "/Users/jdouglass/workspace/natcap/invest/doc/api-docs/conf.py", line 36, in <module>
    for path in os.listdir(INVEST_BUILD_DIR):
FileNotFoundError: [Errno 2] No such file or directory: '/Users/jdouglass/workspace/natcap/invest/doc/api-docs/../../build'

make: *** [dist/apidocs] Error 2
```

So let's follow on this thread within the activated environment:
* `mkdir build` to ensure `build` exists
* `make apidocs`
* Observe error:
```python
make: Circular dist/apidocs <- dist/apidocs dependency dropped.
# -a: always build all files
python3 -m sphinx -a -b html \
		-d build/sphinx/apidocs \
		doc/api-docs build/sphinx/apidocs
Running Sphinx v4.0.1

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "/Users/jdouglass/workspace/natcap/invest/env38/lib/python3.8/site-packages/sphinx/config.py", line 323, in eval_config_file
    exec(code, namespace)
  File "/Users/jdouglass/workspace/natcap/invest/doc/api-docs/conf.py", line 187, in <module>
    INVEST_LIB_DIR
NameError: name 'INVEST_LIB_DIR' is not defined

make: *** [dist/apidocs] Error 2
```

--- a/Makefile
+++ b/Makefile
@@ -188,6 +188,7 @@ clean:
 	-$(RMDIR) $(BUILD_DIR)
 	-$(RMDIR) natcap.invest.egg-info
 	-$(RMDIR) cover
+	-$(RMDIR) doc/api-docs/api
 	-$(RM) coverage.xml
 
 purge: clean
@@ -272,11 +273,12 @@ $(INVEST_BINARIES_DIR): | $(DIST_DIR) $(BUILD_DIR)
 # Documentation.
 # API docs are built in build/sphinx and copied to dist/apidocs
 apidocs: $(APIDOCS_TARGET_DIR) $(APIDOCS_ZIP_FILE)
-$(APIDOCS_TARGET_DIR): | $(DIST_DIR)
+
+$(APIDOCS_BUILD_DIR): install  # need contents of build/lib to build apidocs
 	# -a: always build all files
-	$(PYTHON) -m sphinx -a -b html \
-		-d $(APIDOCS_BUILD_DIR) \
-		doc/api-docs $(APIDOCS_BUILD_DIR)
+	$(PYTHON) -m sphinx -a -b html -d $(APIDOCS_BUILD_DIR)/doctrees doc/api-docs $(APIDOCS_BUILD_DIR)/html
+
+$(APIDOCS_TARGET_DIR): $(APIDOCS_BUILD_DIR) | $(DIST_DIR)
 	# only copy over the built html files, not the doctrees
 	$(COPYDIR) $(APIDOCS_BUILD_DIR)/html $(APIDOCS_TARGET_DIR)
 

--- a/doc/api-docs/conf.py
+++ b/doc/api-docs/conf.py
@@ -37,6 +37,8 @@ for path in os.listdir(INVEST_BUILD_DIR):
     if path.startswith('lib'):
         INVEST_LIB_DIR = os.path.join(INVEST_BUILD_DIR, path)
         break
+else:
+    raise ValueError(f'lib directory not found in {INVEST_BUILD_DIR}')
 
 # -- General configuration ------------------------------------------------
 
@@ -267,15 +269,10 @@ for _loader, name, _is_pkg in itertools.chain(
 # Write sphinx autodoc function for each entrypoint
 with open(MODEL_ENTRYPOINTS_FILE, 'w') as models_rst:
     models_rst.write(MODEL_RST_TEMPLATE)
-    for name, module_title in sorted(all_modules.items(),
-                                     key=lambda x: x[1]):
-        models_rst.write((
-            '{module_title}\n'
-            '{underline}\n'
-            '.. autofunction:: {modname}.execute\n'
-            '   :noindex:\n\n').format(
-                module_title=module_title,
-                underline=''.join(['=']*len(module_title)),
-                modname=name
-            )
-        )
+    for name, module_title in sorted(all_modules.items(), key=lambda x: x[1]):
+        underline = ''.join(['=']*len(module_title))
+        models_rst.write(
+            f'{module_title}\n'
+            f'{underline}\n'
+            f'.. autofunction:: {name}.execute\n'
+            '   :noindex:\n\n')

--- a/doc/api-docs/conf.py
+++ b/doc/api-docs/conf.py
@@ -186,7 +186,8 @@ apidoc.main([
     '--templatedir', os.path.join(DOCS_SOURCE_DIR, 'templates'),  # use custom templates
     '--separate',  # make a separate page for each module
     '--no-toc',  # table of contents page is redundant
-    INVEST_LIB_DIR
+    INVEST_LIB_DIR,
+    '/*.so'  # must be absolute path, see https://github.com/sphinx-doc/sphinx/issues/10200
 ])
 
 

